api/types/resource_ids.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"encoding/json"
+
+	"github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/trace"
+)
+
+// ResourceIDsToString marshals a list of ResourceIDs to a string.
+func ResourceIDsToString(ids []ResourceID) (string, error) {
+	if len(ids) == 0 {
+		return "", nil
+	}
+	bytes, err := json.Marshal(ids)
+	if err != nil {
+		return "", trace.BadParameter("failed to marshal resource IDs to JSON: %v", err)
+	}
+	return string(bytes), nil
+}
+
+// ResourceIDsFromString parses a list for resource IDs from a string. The string
+// should have been obtained from ResourceIDsToString.
+func ResourceIDsFromString(raw string) ([]ResourceID, error) {
+	if raw == "" {
+		return nil, nil
+	}
+	resourceIDs := []ResourceID{}
+	if err := json.Unmarshal([]byte(raw), &resourceIDs); err != nil {
+		return nil, trace.BadParameter("failed to parse resource IDs from JSON: %v", err)
+	}
+	return resourceIDs, nil
+}
```

Comment:
I noticed that this is different from what is in the RFD:

```
tsh request create --resources "db:388aff7f-459f-4a43-804a-3729854976ab,node:3be2fdad-7c79-4cfa-924e-ec1ea7225320"
```

I wish it still followed the format from the RFD because I feel like UX of it is much better than that of JSON objects. You could concatenate those IDs by hand, copying IDs from potentially different sources (other commands or the Web UI if IDs were exposed there). Whereas with JSON objects the user is pretty much forced to copy the output of the command. 😕

But I bet it was necessary to include the cluster name there? Do we have some other commands in tsh/tctl where we pass JSON objects around?

Teleport Connect has its own internal concept of URIs, so 
```
{
"cluster":"enterprise-local",
"kind":"node",
"name":"339cc7f2-2d53-458a-92e2-09e96798c64d"
}
```

becomes `/clusters/enterprise-local/servers/339cc7f2-2d53-458a-92e2-09e96798c64d`.

I suspect might be too late to consider this for this RFD, but I wonder if long term it'd be better for us and the users to implement something like that in our whole codebase.

Reply:
Yeah the cluster name and the resource kind ended up being necessary, I'm not too happy about the current UX. I like the look of `/clusters/enterprise-local/servers/339cc7f2-2d53-458a-92e2-09e96798c64d` though, that seems like it would be much better. Can you point me to any code we have for those URIs?

Reply:
They are [produced by the tsh daemon](https://github.com/gravitational/teleport/blob/2fbf94bf5021b3355aabbed1519005e67c40bdde/lib/teleterm/api/uri/uri.go) and [consumed by the Electron app](https://github.com/gravitational/webapps/blob/627dea4ad701be3c5fae36e06ee73da4837d855c/packages/teleterm/src/ui/uri.ts).

So far tsh daemon was able to get away with not extracting particular identifiers from the URI, but in a more complex scenario you'd probably want the URI producer to be able to do this as well. The Electron app [does extract those identifiers](https://github.com/gravitational/webapps/blob/627dea4ad701be3c5fae36e06ee73da4837d855c/packages/teleterm/src/ui/uri.ts#L20-L30).

Also, I hate to drop this mess onto you, but: so far Teleport Connect conflated the cluster name with the profile name. You probably don't want to dive into [the specifics of it](https://github.com/gravitational/webapps.e/issues/261) at this point. But the summary of them is that in Connect, a URI in the form of `/clusters/foo` should point to the root cluster of profile `foo`. This essentially means that a root cluster name cannot be derived from the URI alone, which I'm not sure is a good trait to have. But for Connect that was okay so far.

Reply:
Thanks for filling me in, I don't want to add too much here so I'll think about this some more and try to improve this in a following PR (the `ResourceIDs(To|From)String` functions here are really just moved from `lib/services/access_request.go`)

Reply:
Sure thing, I think adding URIs to Teleport is a huge ordeal that goes well out of scope of this PR.